### PR TITLE
feat(konnect): assign KongConsumers to KongConsumerGroups

### LINKED ISSUE #519: Konnect entities: Consumer's group assignment
### Problem statement

After #436 and #514 are implemented we can now implement assigning consumers to a group.

### Proposed solution

- Request declarative group assignment API
  - https://kongstrong.slack.com/archives/C03LRB400TC/p1724411291389629
- Use `Consumer`'s group field: https://github.com/Kong/kubernetes-configuration/blob/9ac1d3cf7906e01a6bf14979fef9418195b2053e/api/configuration/v1/kongconsumer_types.go#L56-L59 for assigned groups

### Acceptance criteria

- [ ] KGO can assign `Consumer`s to `ConsumerGroups` in Konnect based on the groups field

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Extends the `KongConsumer` reconciler to handle `KongConsumerGroups` that a given `KongConsumer` is assigned to. Adds envtests suite covering the reconciler expected behavior.

Marked with `do not merge` as it uses a not-released Konnect Go SDK version that relies on a yet-not-merged Konnect API definition PR: https://github.com/Kong/platform-api/pull/724

**Which issue this PR fixes**

Fixes https://github.com/Kong/gateway-operator/issues/519.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: controller/konnect/ops/ops_kongconsumer.go [hunk 1]
```diff
@@ -118,6 +141,218 @@ func updateConsumer(
...
+	cpID string,
+) error {
+	// Resolve the Konnect IDs of the ConsumerGroups referenced by the KongConsumer.
+	desiredConsumerGroupsIDs, invalidConsumerGroups, err := resolveConsumerGroupsKonnectIDs(ctx, consumer, cl)
+	if err != nil {
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				conditions.KonnectEntityProgrammedConditionType,
+				metav1.ConditionFalse,
+				"FailedToResolveConsumerGroupRefs",
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go]: Shouldn't this be a const? 🤔

File: test/envtest/konnect_entities_gatewaycontrolplane_test.go
Comment (pmalek) [inline test/envtest/konnect_entities_gatewaycontrolplane_test.go:1]: I'd feel a bit more comfortable reviewing this PR if we extracted this type of changes to a separate PR. WDYT?

File: controller/konnect/ops/ops_kongconsumer.go [hunk 2]
```diff
@@ -118,6 +141,218 @@ func updateConsumer(
...
+			return errWrapped
+		}
+	}
+
+	return nil
+}
+
+func populateConsumerGroupRefsValidCondition(invalidConsumerGroups []invalidConsumerGroupRef, consumer *configurationv1.KongConsumer) {
+	if len(invalidConsumerGroups) > 0 {
+		var reasons []string
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go]: Preallocate

File: controller/konnect/ops/ops_kongconsumer.go [hunk 3]
```diff
@@ -118,6 +141,218 @@ func updateConsumer(
...
+			RequestBody: &sdkkonnectops.AddConsumerToGroupRequestBody{
+				ConsumerID: lo.ToPtr(consumer.GetKonnectStatus().GetKonnectID()),
+			},
+		})
+		if errWrapped := wrapErrIfKonnectOpFailed(err, UpdateOp, consumer); errWrapped != nil {
+			k8sutils.SetCondition(
+				k8sutils.NewConditionWithGeneration(
+					conditions.KonnectEntityProgrammedConditionType,
+					metav1.ConditionFalse,
+					"FailedToAddToConsumerGroup",
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go]: Const with a comment?

File: controller/konnect/ops/ops_kongconsumer.go [hunk 4]
```diff
@@ -118,6 +141,218 @@ func updateConsumer(
...
+				metav1.ConditionFalse,
+				"FailedToListConsumerGroupsForConsumer",
+				errWrapped.Error(),
+				consumer.GetGeneration(),
+			),
+			consumer,
+		)
+		return errWrapped
+	}
+	// Filter out empty IDs with lo.Compact just in case Konnect returns them.
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go]: Hm, interesting. Why can this happen?

File: controller/konnect/ops/ops_kongconsumer.go [hunk 5]
```diff
@@ -118,6 +141,218 @@ func updateConsumer(
...
+	populateConsumerGroupRefsValidCondition(invalidConsumerGroups, consumer)
+
+	// Reconcile the ConsumerGroups assigned to the KongConsumer in Konnect (list the actual ConsumerGroups, calculate the
+	// difference, and add/remove the Consumer from the ConsumerGroups accordingly).
+	if err := reconcileConsumerGroupsWithKonnect(ctx, desiredConsumerGroupsIDs, cgSDK, cpID, consumer); err != nil {
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				conditions.KonnectEntityProgrammedConditionType,
+				metav1.ConditionFalse,
+				"FailedToReconcileConsumerGroupsWithKonnect",
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go]: Const with a comment

**Review by pmalek — COMMENTED** [review]
Looks solid (given the API limitations in Konnect)

Left some comments for consideration.

File: test/envtest/konnect_entities_gatewaycontrolplane_test.go
Comment (czeslavo) [reply test/envtest/konnect_entities_gatewaycontrolplane_test.go:1]: 👍 PTAL https://github.com/Kong/gateway-operator/pull/607

File: controller/konnect/ops/ops_kongconsumer.go (hunk 4 above)
Comment (czeslavo) [reply controller/konnect/ops/ops_kongconsumer.go]: It can _technically_ happen because the `sdkkonnectcomp.ConsumerGroup` in the response has `ID *string` field. Below I'm transforming all of those into a slice of strings and using `lo.Compat` I'm dropping the empty ones.

**Comment by czeslavo** [comment]
@pmalek Your comments should be addressed now: 21de6d4. PTAL

File: controller/konnect/ops/ops_kongconsumer.go [hunk 6]
```diff
@@ -118,6 +141,189 @@ func updateConsumer(
...
+	cpID string,
+	consumer *configurationv1.KongConsumer,
+) error {
+	// List the ConsumerGroups that the Consumer is assigned to in Konnect.
+	cgsResp, err := cgSDK.ListConsumerGroupsForConsumer(ctx, sdkkonnectops.ListConsumerGroupsForConsumerRequest{
+		ControlPlaneID: cpID,
+		ConsumerID:     consumer.GetKonnectStatus().GetKonnectID(),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list ConsumerGroups for Consumer %s/%s: %w", consumer.GetNamespace(), consumer.GetName(), err)
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go]: Instead of getting the ns and name you can use 

```suggestion
		return fmt.Errorf("failed to list ConsumerGroups for Consumer %s: %w", client.ObjectKeyFromObject(consumer), err)
```

**Review by programmer04 — APPROVED** [review]
LGTM

File: test/integration/test_konnect_entities.go [hunk 7]
```diff
@@ -214,13 +170,35 @@ func TestKonnectEntities(t *testing.T) {
...
+		},
+		Spec: configurationv1.KongConsumerSpec{
+			ControlPlaneRef: &configurationv1alpha1.ControlPlaneRef{
+				Type:                 configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef,
+				KonnectNamespacedRef: &configurationv1alpha1.KonnectNamespacedRef{Name: cp.Name},
+			},
+		},
+	}
+	err = GetClients().MgrClient.Create(GetCtx(), &kc)
+	require.NoError(t, err)
```
Comment (pmalek) [inline test/integration/test_konnect_entities.go]:
```suggestion
	require.NoError(t, GetClients().MgrClient.Create(GetCtx(), &kc))
```

File: controller/konnect/ops/ops_kongconsumer.go [hunk 8]
```diff
@@ -118,6 +141,192 @@ func updateConsumer(
 	return nil
 }
 
+// handleConsumerGroupAssignments resolves ConsumerGroup references of a KongConsumer, reconciles them with Konnect, and
```
Comment (pmalek) [inline controller/konnect/ops/ops_kongconsumer.go:144]: Given how many round trips this is going to be and the fact that we have an issue open for declarative consumer group membership API: KOKO-1952, can we add a comment somewhere that this approach - listing, adding and removing - is a stopgap before we get KOKO-1952 implemented?

**Review by pmalek — COMMENTED** [review]
Looks good, just 2 minor nits.

File: controller/konnect/ops/ops_kongconsumer.go (hunk 8 above)
Comment (czeslavo) [reply controller/konnect/ops/ops_kongconsumer.go:144]: Good call - added a tracking issue for that as well: https://github.com/Kong/gateway-operator/issues/634

**Review by pmalek — APPROVED** [review]
👍 

I'd consider adding the group assignment in the sample that we have for consumer groups: https://github.com/Kong/gateway-operator/blob/38fcef4ddde9b1d24ff223f94ebada26cb400f9f/config/samples/konnect_kongconsumergroup.yaml 

WDYT?

**Review by pmalek — APPROVED** [review]

**Review by pmalek — APPROVED** [review]
